mlir/lib/Dialect/Affine/IR/AffineOps.cpp

```diff
@@ -2454,27 +2454,25 @@ bool AffineForOp::matchingBoundOperandList() {
 
 SmallVector<Region *> AffineForOp::getLoopRegions() { return {&getRegion()}; }
 
-std::optional<Value> AffineForOp::getSingleInductionVar() {
-  return getInductionVar();
-}
+ValueRange AffineForOp::getInductionVars() { return {getInductionVar()}; }
 
-std::optional<OpFoldResult> AffineForOp::getSingleLowerBound() {
+SmallVector<OpFoldResult> AffineForOp::getMixedLowerBound() {
   if (!hasConstantLowerBound())
-    return std::nullopt;
+    return {};
   OpBuilder b(getContext());
-  return OpFoldResult(b.getI64IntegerAttr(getConstantLowerBound()));
+  return {OpFoldResult(b.getI64IntegerAttr(getConstantLowerBound()))};
 }
```

Comment:
i have a feeling this won't fly, so i'm happy to hear thoughts

Reply:
Can you explain a little bit as to why the semantics are not quite right?

Reply:
Its supposed to return mixed values, but here i'm only returning a non empty vector if the result is a constant. i did this to keep the same semantics for the older interface methods on the affine::ForOp. but really it should return any OpFoldResult. atm this doesn't affect anything because this op doesn't use the new methods anyway

Reply:
my issue was that i didn't really fully understand how affine loop bounds work compared to scf. because of the whole `AffineBound` thing, which i'm unfamiliar with

Reply:
I dont know if you need to worry about it? the induction variable is just an SSA value.

Reply:
i'm refering to the bounds and steps

Reply:
Yeah, I dont know either. @bondhugula, @Groverkss or @stephenneuendorffer   can you guys comment?

Reply:
The bound is computed as an (affine) function of multiple values, which are passed as operands to the loop operation. It cannot be represented with one value since the result of that function is not materialized. All existing interface methods are "best effort", i.e., they return non-nullopt value if it makes sense for the specific loop. And should be documented as such if they aren't already. So it is fine to return nullopt for affine loops with non-constant upper bound.

If we _really_ needed that, we could add a `materializeLower/UpperBound(OpBuilder &)` method, but I doubt it's useful.